src/util/scratch-link-websocket.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * This class provides a ScratchLinkSocket implementation using WebSockets,
+ * attempting to connect with the locally installed Scratch-Link.
+ *
+ * To connect with ScratchLink without WebSockets, you must implement all of the
+ * public methods in this class.
+ * - open()
+ * - close()
+ * - set[Open|Close|Error]
+ * - setHandleMessage
+ * - isOpen()
+ */
+class ScratchLinkWebSocket {
+    constructor (type) {
+        this._type = type;
+        this._onOpen = null;
+        this._onClose = null;
+        this._onError = null;
+        this._handleMessage = null;
+
+        this._ws = null;
+    }
+
+    open () {
+        switch (this._type) {
+        case 'BLE':
+            this._ws = new WebSocket('wss://device-manager.scratch.mit.edu:20110/scratch/ble');
+            break;
+        case 'BT':
+            this._ws = new WebSocket('wss://device-manager.scratch.mit.edu:20110/scratch/bt');
+            break;
+        default:
+            throw new Error(`Unknown ScratchLink socket Type: ${this._type}`);
+        }
+
+        if (this._onOpen && this._onClose && this._onError) {
+            this._ws.onopen = this._onOpen;
+            this._ws.onclose = this._onClose;
+            this._ws.onerror = this._onError;
+        } else {
+            throw new Error('Must set open, close and error handlers before calling open on the socket');
```

Comment:
Obviously this error should never been seen by a user, so it's not something we should worry about localizing. Would it be worth having a convention for this type of error to be clear that it's not something the end user would see? e.g. `Error('INTERNAL: Must set...')`

Just started thinking about this after reading the API endpoint guide: https://github.com/LLK/scratch-api/wiki/Endpoint-route-design-guide#responses-should-be-machine-readable